packages/coding-standards/Jetpack/Sniffs/InlineDocs/HooksMustHaveDocblockSniff.php

```diff
@@ -0,0 +1,157 @@
+<?php
+/**
+ * This sniff verifies a WordPress hook function has a preceding docblock.
+ *
+ * @package   automattic/jetpack-coding-standards
+ */
+
+namespace Automattic\Jetpack\CodingStandards\Sniffs\InlineDocs;
+
+use WordPressCS\WordPress\AbstractFunctionRestrictionsSniff;
+use PHP_CodeSniffer\Util\Tokens;
+
+/**
+ * Class HooksMustHaveDocblockSniff
+ *
+ * @package Automattic\Jetpack\CodingStandards\Sniffs\InlineDocs
+ */
+class HooksMustHaveDocblockSniff extends AbstractFunctionRestrictionsSniff {
+
+	/**
+	 * Array of WordPress hook execution functions.
+	 *
+	 * @var array WordPress hook execution function name => filter or action.
+	 */
+	protected $hook_functions = array(
+		'apply_filters'            => 'filter',
+		'apply_filters_ref_array'  => 'filter',
+		'apply_filters_deprecated' => 'filter',
+		'do_action'                => 'action',
+		'do_action_ref_array'      => 'action',
+		'do_action_deprecated'     => 'action',
+	);
+
+	/**
+	 * Groups of functions to restrict.
+	 *
+	 * Example: groups => array(
+	 *  'lambda' => array(
+	 *      'type'      => 'error' | 'warning',
+	 *      'message'   => 'Use anonymous functions instead please!',
+	 *      'functions' => array( 'file_get_contents', 'create_function' ),
+	 *  )
+	 * )
+	 *
+	 * @return array
+	 */
+	public function getGroups() {
+		return array(
+			'hooks' => array(
+				'functions' => array_keys( $this->hook_functions ),
+			),
+		);
+	}
+
+	/**
+	 * Process a matched token.
+	 *
+	 * @since 1.0.0 Logic split off from the `process_token()` method.
+	 *
+	 * @param int    $stack_ptr       The position of the current token in the stack.
+	 * @param string $group_name      The name of the group which was matched.
+	 * @param string $matched_content The token content (function name) which was matched.
+	 *
+	 * @return int|void Integer stack pointer to skip forward or void to continue
+	 *                  normal file processing.
+	 */
+	public function process_matched_token( $stack_ptr, $group_name, $matched_content ) {
+
+		$func_open_paren_token = $this->phpcsFile->findNext( Tokens::$emptyTokens, ( $stack_ptr + 1 ), null, true );
+		if ( false === $func_open_paren_token
+			|| \T_OPEN_PARENTHESIS !== $this->tokens[ $func_open_paren_token ]['code']
+			|| ! isset( $this->tokens[ $func_open_paren_token ]['parenthesis_closer'] )
+		) {
+			// Live coding, parse error or not a function call.
+			return;
+		}
+
+		$previous_comment = $this->phpcsFile->findPrevious( Tokens::$commentTokens, ( $stack_ptr - 1 ) );
+
+		if ( false !== $previous_comment ) {
+			/*
+			 * Check to determine if there is a comment immediately preceding the function call.
+			 */
+			if ( ( $this->tokens[ $previous_comment ]['line'] + 1 ) !== $this->tokens[ $stack_ptr ]['line'] ) {
```

Comment:
Should double-check whether DocBlock association needs to be on the immediate line before, or just the thing before it when whitespace is excluded. i.e.:

```php
/**
 * ...
 */
apply_filters( ... );
```

```php
/**
 * ...
 */

apply_filters( ... );
```

Both may be correctly associated. It should have parity with the [(draft) PSR-5](https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc.md#3-definitions) (my emphasis):

> It is RECOMMENDED to precede a "Structural Element" with a DocBlock where it is defined and not with each usage. It is common practice to have the DocBlock precede a Structural Element but **it MAY also be separated by an undetermined number of empty lines**.